Source/Core/DolphinQt2/Config/LoggerWidget.cpp

```diff
@@ -0,0 +1,468 @@
+      {
+        LogManager::GetInstance()->AddListener(log_type, LogListener::LOG_WINDOW_LISTENER);
+      }
+      else
+      {
+        LogManager::GetInstance()->RemoveListener(log_type, LogListener::LOG_WINDOW_LISTENER);
+      }
+    }
+  }
+
+  m_ini_options->Set("WriteToFile", m_out_file->isChecked());
+  m_ini_options->Set("WriteToConsole", m_out_console->isChecked());
+  m_ini_options->Set("WriteToWindow", m_out_window->isChecked());
+
+  // Config - Log Types
+  for (int i = 0; i < LogTypes::NUMBER_OF_LOGS; ++i)
+  {
+    const auto type = static_cast<LogTypes::LOG_TYPE>(i);
+    const auto name = LogManager::GetInstance()->GetShortName(type);
+    bool enabled = m_types_list->item(i)->checkState() == Qt::Checked;
+    bool was_enabled;
+    m_ini_logs->Get(name, &was_enabled, false);
+
+    if (enabled != was_enabled)
+    {
+      LogManager::GetInstance()->SetEnable(type, enabled);
+      m_ini_logs->Set(name, enabled);
+    }
+  }
+
+  // Log - Wrap Lines
+  m_ini_options->Set("WrapLines", m_log_wrap->isChecked());
+  m_log_text->setLineWrapMode(m_log_wrap->isChecked() ? QTextEdit::WidgetWidth : QTextEdit::NoWrap);
+
+  // Log - Font Selection
+  m_ini_options->Set("Font", m_log_font->currentIndex());
+  UpdateFont();
+
+  // Window tracking
+  // Note: Option names are not capitalized to be consistent with Wx
+  auto position = pos();
+
+  m_ini_logwindow->Set("floating", isFloating());
+  m_ini_logwindow->Set("x", position.x());
+  m_ini_logwindow->Set("y", position.y());
+  m_ini_logwindow->Set("width", width());
+  m_ini_logwindow->Set("height", height());
+
+  m_ini.Save(File::GetUserPath(F_LOGGERCONFIG_IDX));
+}
+
+void LoggerWidget::OnTabVisibilityChanged()
+{
+  bool log_visible = SConfig::GetInstance().m_InterfaceLogWindow;
+  bool config_visible = SConfig::GetInstance().m_InterfaceLogConfigWindow;
+
+  m_tab_widget->clear();
+  if (log_visible)
+    m_tab_widget->addTab(m_tab_log, tr("Log"));
+
+  if (config_visible)
+    m_tab_widget->addTab(m_tab_config, tr("Configuration"));
+
+  if (!log_visible && !config_visible)
+    hide();
+  else
+    show();
+}
+
+void LoggerWidget::Log(LogTypes::LOG_LEVELS level, const char* text)
+{
+  std::lock_guard<std::mutex> lock(m_log_mutex);
+
+  const char* color = "white";
+
+  switch (level)
+  {
+  case LogTypes::LOG_LEVELS::LERROR:
+    color = "red";
+    break;
+  case LogTypes::LOG_LEVELS::LWARNING:
+    color = "yellow";
+    break;
+  case LogTypes::LOG_LEVELS::LNOTICE:
+    color = "green";
+    break;
+  case LogTypes::LOG_LEVELS::LINFO:
+    color = "cyan";
+    break;
+  case LogTypes::LOG_LEVELS::LDEBUG:
+    color = "lightgrey";
+    break;
+  }
+
+  m_log_queue.push(QStringLiteral("%1 <font color='%2'>%3</font>")
+                       .arg(QString::fromStdString(std::string(text).substr(0, 10)),
+                            QString::fromStdString(color),
+                            QString::fromStdString(std::string(text).substr(10))));
```

Comment:
How guaranteed is it that our timestamp always has 10 characters? Plus, maybe leave a note or something that the `10` in there refers to the timestamp in front of the message, otherwise it seems too arbitrary.